nemo/collections/nlp/modules/common/megatron/retrieval_service.py

```diff
@@ -97,3 +207,272 @@ def get_knn(self, query: Union[List[str], str, torch.Tensor]):
             # unpack the single sentence input
             return results[0]
         return np.stack(results, axis=0).astype(np.int64)
+
+
+class RetrievalServer(object):
+    def __init__(
+        self, faiss_index: str, faiss_devices: str, nprobe: int, retrieval_index: str, tokenizer: TokenizerSpec,
+    ):
+        self.app = Flask(__name__, static_url_path='')
+        # server
+        has_gpu = torch.cuda.is_available() and hasattr(faiss, "index_gpu_to_cpu")
+
+        if faiss_devices is None or not torch.cuda.is_available():
+            device_list = None
+        else:
+            device_list = ['cuda:' + str(device) for device in faiss_devices.split(',')]
+
+        self.index = faiss.read_index(faiss_index)
+        if has_gpu and device_list is not None:
+            beg = time.time()
+            co = faiss.GpuMultipleClonerOptions()
+            co.useFloat16 = True
+            co.usePrecomputed = False
+            co.shard = True
+            self.index = faiss.index_cpu_to_all_gpus(self.index, co, ngpu=len(device_list))
+            end = time.time()
+            logging.info('convert Faiss db to GPU takes', end - beg)
+        self.index.nprobe = nprobe
+        self.tokenizer = tokenizer
+        self.ds = MMapRetrievalIndexedDataset(retrieval_index)
+        api = Api(self.app)
+        api.add_resource(
+            FaissRetrievalResource, '/knn', resource_class_args=[self.index, self.tokenizer, self.ds,],
+        )
+
+    def run(self, url, port=PORT_NUM):
+        threading.Thread(target=lambda: self.app.run(host=url, threaded=True, port=port)).start()
+
+
+class DynamicRetrievalResource(FaissRetrievalResource):
```

Comment:
What's the use case for this over plain `RetrievalResource`? A few comments would be helpful here imo.